samples/go/photo-dedup/job/hashjob.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package job
+
+import (
+	"fmt"
+	"image"
+	_ "image/gif"
+	_ "image/jpeg"
+	_ "image/png"
+	"math"
+	"net/http"
+	"os"
+	"runtime"
+	"sync"
+
+	"github.com/attic-labs/noms/go/datas"
+	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/noms/go/util/status"
+	"github.com/attic-labs/noms/go/walk"
+	"github.com/attic-labs/noms/samples/go/photo-dedup/dhash"
+	"github.com/attic-labs/noms/samples/go/photo-dedup/model"
+)
+
+// HashPhotosJob adds a dhash field to every photo in photoSets and commits them to
+// a new Photo set in outDS. The dhash is used in turn by the DeduplicatePhotosJob
+// to group similar photos into PhotoGroups.
+func HashPhotosJob(db datas.Database, photoSets []types.Value, outDS datas.Dataset) error {
+	return commitHashedPhotos(db, outDS, hashPhotos(db, photoSets))
+}
+
+// hashPhotos adds a dhash to each photo in photoSets and delivers them on the returned channel
+func hashPhotos(db datas.Database, photoSets []types.Value) <-chan types.Struct {
+	photosToHashCh := make(chan model.Photo)
+	hashedPhotosCh := make(chan types.Struct)
+
+	go func() {
+		for _, set := range photoSets {
+			walk.WalkValues(set, db, func(cv types.Value) (stop bool) {
+				if photo, ok := model.UnmarshalPhoto(cv); ok {
+					photosToHashCh <- photo
+				}
+				return false
+			})
+		}
+		close(photosToHashCh)
+	}()
+
+	fmt.Print("Downloading and hashing photos...\n")
+	numWorkers := runtime.NumCPU() * 4
+	var wg sync.WaitGroup
+	wg.Add(numWorkers)
+	for i := 0; i < numWorkers; i++ {
+		go func() {
+			defer wg.Done()
+			for photo := range photosToHashCh {
+				hashed, err := addHashToPhoto(photo)
+				if err != nil {
+					fmt.Fprintf(os.Stderr, "\nSkipping: %s\n", err)
+					hashedPhotosCh <- photo.Marshal()
+				} else {
+					hashedPhotosCh <- hashed.Marshal()
+				}
+			}
+		}()
+	}
+	go func() {
+		wg.Wait()
+		fmt.Print("\nHashing complete\n")
+		close(hashedPhotosCh)
+	}()
+	return hashedPhotosCh
+}
+
+func addHashToPhoto(photo model.Photo) (model.Photo, error) {
+	url := pickBestImage(photo)
+	if url == "" {
+		return nil, fmt.Errorf("No URL found for photo %s", photo.ID())
+	}
+	res, err := http.Get(string(url))
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+	img, _, err := image.Decode(res.Body)
+	if err != nil {
+		if err == image.ErrFormat {
+			err = fmt.Errorf("%s: unknown format", url)
+		}
+		return nil, err
+	}
+	photo.SetDhash(dhash.New(img))
+	return photo, nil
+}
+
+// pickBestImage returns the image URL corresponding to the size
+// closest to but not below the ideal size. If there is no image
+// large enough, it returns the next smallest image.
+func pickBestImage(photo model.Photo) string {
```

Comment:
This should be factored out somewhere ... I feel like we have it, or something like it, in several places.